torch/_inductor/runtime/benchmarking.py

```diff
@@ -0,0 +1,1001 @@
+
+
+class LazyBenchmark:
+    def __init__(self: Self, benchmark: Callable[[], float]) -> None:
+        self.benchmark = benchmark
+
+    @cached_property
+    def timing_ms(self: Self) -> float:
+        counters["inductor"]["benchmarking_finalize_lazy_benchmark"] += 1
+        timing_ms = self.benchmark()
+        # I don't think this helps with saving memory at all,
+        # but at least it gives good signal if we ever try
+        # to call self.benchmark again
+        del self.benchmark
+        return timing_ms
+
+    def __float__(self: Self) -> float:
+        return float(self.timing_ms)
+
+    def __format__(self: Self, format_spec: str) -> str:
+        return format(self.timing_ms, format_spec)
+
+    def __str__(self: Self) -> str:
+        return str(self.timing_ms)
+
+    def __lt__(self: Self, other: Any) -> bool:
+        return self.timing_ms < other
+
+    def __le__(self: Self, other: Any) -> bool:
+        return self.timing_ms <= other
+
+    def __gt__(self: Self, other: Any) -> bool:
+        return self.timing_ms > other
+
+    def __ge__(self: Self, other: Any) -> bool:
+        return self.timing_ms >= other
+
+    def __add__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return self.timing_ms + other
+        return LazyBenchmark(lambda: self.timing_ms + other)
+
+    def __radd__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return other + self.timing_ms
+        return LazyBenchmark(lambda: other + self.timing_ms)
+
+    def __sub__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return self.timing_ms - other
+        return LazyBenchmark(lambda: self.timing_ms - other)
+
+    def __rsub__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return other - self.timing_ms
+        return LazyBenchmark(lambda: other - self.timing_ms)
+
+    def __mul__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return self.timing_ms * other
+        return LazyBenchmark(lambda: self.timing_ms * other)
+
+    def __rmul__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return other * self.timing_ms
+        return LazyBenchmark(lambda: other * self.timing_ms)
+
+    def __truediv__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return self.timing_ms / other
+        return LazyBenchmark(lambda: self.timing_ms / other)
+
+    def __rtruediv__(self: Self, other: Any) -> Any:
+        if not hasattr(self, "benchmark"):
+            return other / self.timing_ms
+        return LazyBenchmark(lambda: other / self.timing_ms)
+
+
+def is_cpu_device(inputs: List[Any]) -> bool:
+    return all(
+        _input.device == torch.device("cpu")
+        for _input in inputs
+        if isinstance(_input, torch.Tensor)
+    )
+
+
+class Benchmarker:
```

Comment:
Do you think it make sense to have both an "online" and an "offline" benchmarker that works in different scenarios? For example, the offline benchmarker has a more relaxed constraint of benchmarking time (like using cuda graph)

Reply:
My guess is that using cudagraphs to benchmark wouldn't actually help with performance enough to warrant the complexity and reliability concerns that would come from trying to offline the benchmarking, if it even helped with performance at all.

The one concrete thing we get from using cudagraphs to benchmark is the ability to exclude GPU launch overhead from our measurements, but I'm not convinced this matters is 99% of cases. The most basic use of a benchmarker is autotuning a kernel to one set of optimal configs, for which GPU launch overhead is basically irrelevant since it is fairly stable and should affect each config in an equal and linear fashion (so we still get the same ordering of configs, which is all that is important). The singular case I can think of when GPU launch overhead might play a factor in actual E2E performance is when we are comparing fusion opportunities. If we want to try fusing some kernels A + B into a new C kernel we first have to check that the runtime of C is less than the runtime of A + B to know if it is a "good" fusion, the problem here is that we are measuring the GPU launch overhead twice when benchmarking A + B and only once when benchmarking C. But then again, GPU launch overhead is only 5us, so I'm not sure this even has a huge impact.

One way to know for sure is to just do a performance run that always uses cudagraphs for benchmarking, so we'll see how that goes (https://github.com/pytorch/pytorch/actions/runs/10184709063).

Reply:
Yeah I agree that if we can get good enough benchmark results without cudagraph, we should not use it.

I am also wondering do you think `torch._inductor.runtime.benchmarking` is good enough for both online and offline benchmarking? Or do you think we could to do additional stuff in an offline benchmarker, which has relaxed constraints on benchmarking time to enable getting higher quality of benchmarking results?

I have a question regarding to the term "GPU launch overhead", I guess it is different from "GPU kernel launch overhead"? I thought the GPU kernel launch overhead is already excluded even without CUDA graph, since we are measuring the

Reply:
> I am also wondering do you think torch._inductor.runtime.benchmarking is good enough for both online and offline benchmarking?

If it works online it should work offline

> Or do you think we could to do additional stuff in an offline benchmarker, which has relaxed constraints on benchmarking time to enable getting higher quality of benchmarking results?

Maybe, I'd have to think about it. We would have stricter control over the state of the device we are benchmarking on (i.e. for GPU we could pin the frequency however we wanted), and we'd definitely be able to benchmark for a longer duration. Whether either of these would actually make a difference in E2E performance is anyone's guess. The real winner of offline benchmarking imo would be saving compile time by not having to materialize every template, but at that point we're really just talking about offlining autotuning

Reply:
> I have a question regarding to the term "GPU launch overhead", I guess it is different from "GPU kernel launch overhead"? I thought the GPU kernel launch overhead is already excluded even without CUDA graph, since we are measuring the elapsed time between 2 CUDA events?

I think we're talking about the same thing. As for what the cuda events actually record, I believe it does include the kernel launch overhead. My understanding is that when you hit record on our start event it records that instant, then you launch and your kernel gets queued up (hence we incur the launch overhead), it runs, and then we hit record again on our end event.

Reply:
I see, so the "cpu launch overhead" and "gpu kernel launch overhead" are two different metrics. CUDA Graph can eliminate both, while Non-CUDA-Graph can eliminate the first but not the latter (with torch.cuda.Event).

Reply:
> I see, so the "cpu launch overhead" and "gpu kernel launch overhead" are two different metrics. CUDA Graph can eliminate both, while Non-CUDA-Graph can eliminate the first but not the latter (with torch.cuda.Event).

Yes, CPU launch overhead is much more variable (for example, consider the case where our kernel needs to allocate an output tensor at runtime) and GPU kernel launch overhead is pretty much static. Benchmarking with cudagraphs eliminates the GPU kernel launch overhead; as for the CPU launch overhead I'm not sure, it should definitely reduce the total overhead but we still need to eat some of it in between recording our start event and the graph replaying. We can completely eliminate CPU launch overhead when not using cudagraphs to benchmark, but indeed we then have to take the GPU kernel launch overhead.

Reply:
Seems like `do_bench_cudagraphs` is actually worse for performance (and obviously compile time). My best guess is that we really hurt benchmark validity from skipping the cache clear in between executions of the kernel within the graph, and probably also from the lack of a stable warmup cycle before the graph gets replayed.
![Uploading Screenshot 2024-08-01 at 2.15.28 AM.png…]()